tensorboard/data/server/data_compat.rs

```diff
@@ -103,6 +105,51 @@ impl EventValue {
         };
         match *value_box {
             pb::summary::value::Value::Tensor(tp) => Ok(tp),
+            pb::summary::value::Value::Histo(hp) => {
+                // Migrate legacy TF 1.x HistogramProto to TensorProto. The "spec" in summary.proto
+                // says `bucket` and `bucket_limit` are parallel arrays encoding bucket counts and
+                // bucket right edges; the first bucket's left edge is assumed to be -DBL_MAX and
+                // subsequent left edges are defined as the right edge of the preceeding bucket.
+                //
+                // Our conversion logic in data_compat.py however disobeys this and instead sets the
+                // leftmost and rightmost edges to the `min` and `max` values, respectively. This
+                // will result in the outermost buckets having left edge > right edge if they were
+                // originally empty. Apparently the histogram visualization can't handle buckets
+                // extending to -/+ DBL_MAX, but can handle buckets of negative width?
+                //
+                // For consistency with the status quo, we replicate this questionable logic here.
+                let num_buckets = hp.bucket.len();
+                let bucket_edges = || hp.bucket_limit.iter().take(num_buckets - 1).copied();
```

Comment:
Shall we check that `bucket.len() == bucket_limit.len()` and signal data
loss otherwise? or at least take `num_buckets` to be the shorter of the
two lengths? As written, it looks like `bucket_lefts` could be much
shorter than `bucket_counts`, which would make the output shape not
correct.

Also, if `hp.bucket.len() == 0`, the `num_buckets - 1` is an error
(panic in debug mode), so let’s do something nicer in that case?
Maybe worth a test (up to you).

In case you’re interested, my [sketch of this routine][1] took quite a
different approach (imperative instead of streamy). I think that I like
your idea to use `tensor_content`, since at least cloning a `Bytes` is
cheap compared to cloning a `Vec<f64>`.

[1]: https://github.com/wchargin/rustboard-server/blob/05550d7b356b94c7e0dea335030d4d11ab6510b8/src/run.rs#L257

Reply:
Lovely; thanks! Yeah, I’m perfectly happy with the streams here. I’m
part of the “use streams judiciously” crowd, and this looks like a
pretty easy sell. It’s definitely more readable than mine.